src/Libraries/Microsoft.Extensions.TimeProvider.Testing/FakeTimeProvider.cs

```diff
@@ -12,83 +12,109 @@
     /// </summary>
+    /// <param name="startTime">The initial time and date reported by the provider.</param>
     /// <remarks>
-    /// This creates a clock whose time is set to midnight January 1st 2000.
-    /// The clock is set to not automatically advance every time it is read.
+    /// The provider is set to not automatically advance time each time it is read.
     /// </remarks>
-    public FakeTimeProvider()
+    public FakeTimeProvider(DateTimeOffset startTime)
     {
-        _localTimeZone = TimeZoneInfo.Utc;
-        _now = Epoch;
+        _now = startTime;
     }
 
     /// <summary>
     /// Initializes a new instance of the <see cref="FakeTimeProvider"/> class.
     /// </summary>
-    /// <param name="epoch">The starting point for the clock used by this <see cref="FakeTimeProvider"/>.</param>
-    [Experimental]
-    public FakeTimeProvider(DateTimeOffset epoch)
-        : this()
+    /// <param name="startTime">The initial time and date reported by the provider.</param>
+    /// <param name="autoAdvanceAmount">The amount of time by which to automatically advance time whenever it is read.</param>
+    public FakeTimeProvider(DateTimeOffset startTime, TimeSpan autoAdvanceAmount)
     {
-        Epoch = epoch;
-        _now = epoch;
+        _ = Throw.IfLessThan(autoAdvanceAmount.Ticks, 0);
+
+        _now = startTime;
+        _autoAdvanceAmount = autoAdvanceAmount;
     }
 
     /// <inheritdoc />
     public override DateTimeOffset GetUtcNow()
     {
-        return _now;
+        List<FakeTimeProviderTimer.Waiter>? waitersToWake;
+        DateTimeOffset result;
+
+        lock (Waiters)
+        {
+            result = _now;
+            _now += _autoAdvanceAmount;
+            waitersToWake = GetWaitersToWake();
+        }
+
+        WakeWaiters(waitersToWake);
+        return result;
     }
 
     /// <summary>
-    /// Sets the date and time in the UTC timezone.
+    /// Sets the date and time in the UTC time zone.
     /// </summary>
-    /// <param name="value">The date and time in the UTC timezone.</param>
+    /// <param name="value">The date and time in the UTC time zone.</param>
     public void SetUtcNow(DateTimeOffset value)
     {
-        if (value < _now)
-        {
-            Throw.ArgumentOutOfRangeException(nameof(value), $"Cannot go back in time. Current time is {GetUtcNow()}.");
-        }
+        List<FakeTimeProviderTimer.Waiter>? waitersToWake = null;
 
-        while (value >= _now && TryGetWaiterToWake(value) is FakeTimeProviderTimer.Waiter waiter)
+        lock (Waiters)
         {
-            _now = waiter.WakeupTime;
-            waiter.TriggerAndSchedule(false);
+            if (value < _now)
+            {
+                Throw.ArgumentOutOfRangeException(nameof(value), $"Cannot go back in time. Current time is {_now}.");
+            }
+
+            _now = value;
+            waitersToWake = GetWaitersToWake();
         }
 
-        _now = value;
+        WakeWaiters(waitersToWake);
```

Comment:
I think this may break the promise that callbacks happen in the correct order.

E.g., we have two timers:

1. `timer1` has callback `duetime = 1 second` and `period = 1 second`
2. `timer2` has callback `duetime = 3 seconds` and `period = infinite`

If call advance time by 3 seconds, then I expect the callback order to be `[timer1, timer1, timer2]`, but as far as I can tell, the code in the PR will invoke the timers callbacks `[timer1, timer2]` only.

Reply:
Indeed, it doesn't guarantee order. And I think that's fine, there is no promise here. This is fundamentally racy in nature and ordering here is an illusion anyway for those edge cases, as the processor scheduler can readily change the effective order that the waiters are woken up.

Reply:
If two callbacks have the same scheduled callback time, then yes, I agree that the ordering is not likely to be guaranteed in production either, that said, for the sake of deterministic tests, it would be nice if they were.

However, in my example above, timer1's callback should be called twice before timer2 callback. That was one of the points of https://github.com/dotnet/extensions/pull/4022, but I realize I did not include a test with multiple timers.

This test should ensure the correct behaviour (just typed out here, not actually compiled/tested):

```c#
[Fact]
public void Timers_callbacks_are_ordered()
{
    var callbackTimes = new List<(int TimerId, DateTimeOffset CallbackTime)>();
    var timeProvider = new FakeTimeProvider();
    var start = timeProvider.GetUtcNow();
    var oneSecond = TimeSpan.FromSeconds(1);
    var threeSeconds = TimeSpan.FromSeconds(3);
    var timer1 = timeProvider.CreateTimer(_ => { callbackTimes.Add((1, timeProvider.GetUtcNow())); }, null, oneSecond, oneSecond);
    var timer2 = timeProvider.CreateTimer(_ => { callbackTimes.Add((2, timeProvider.GetUtcNow())); }, null, threeSeconds, threeSeconds);

    timeProvider.Advance(threeSeconds);

    Assert.Equal(new[]
    {
      (1, start + oneSecond),
      (1, start + oneSecond + oneSecond),
      (2, start + threeSeconds),
      (1, start + threeSeconds),
    },
    callbackTimes);
}
```

Reply:
I revamped things and ordering promises are back.